src/directwrite/mod.rs

```diff
@@ -238,12 +231,25 @@ impl crate::Rasterize for DirectWriteRasterizer {
     fn get_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
         let loaded_font = self.get_loaded_font(glyph.font_key)?;
 
-        match self.rasterize_glyph(&loaded_font.face, glyph.size, glyph.c) {
-            Err(err @ Error::MissingGlyph(_)) => {
-                let fallback_font = self.get_fallback_font(&loaded_font, glyph.c).ok_or(err)?;
-                self.rasterize_glyph(&fallback_font.create_font_face(), glyph.size, glyph.c)
-            },
-            result => result,
+        let fallback_font;
+        let mut glyph_index = self.get_glyph_index(&loaded_font.face, glyph.character);
+        let font = if glyph_index != MISSING_GLYPH_INDEX {
+            loaded_font
+        } else if let Some(font) = self.get_fallback_font(&loaded_font, glyph.character) {
+            fallback_font = Font::from(font);
+            glyph_index = self.get_glyph_index(&fallback_font.face, glyph.character);
+            &fallback_font
+        } else {
+            loaded_font
+        };
```

Comment:
This code looks a lot more complicated now because there's just way more branching one has to understand. What's the reason for this change? Could you maybe restructure it to use two conditionals at most?

One idea without changing too much code:

```rust
let mut font = &loaded_font;
let mut glyph_index = self.get_glyph_index(&loaded_font.face, glyph.character);
if glyph_index == MISSING_GLYPH_INDEX {
    if let Some(font) = self.get_fallback_font(&loaded_font, glyph_character) {
        glyph_index = self.get_glyph_index(&fallback_font.face, glyph.character);
        fallback_font = Font::from(font);
        font = &fallback_font;
    }
}
```

But I think we can do **way** better than that by using the abstraction magic of methods. Say we move the check for the glyph index in the `get_fallback_font` method. We could easily have an early return if the loaded font already contains the glyph, which would probably clean things up a lot. We also wouldn't need the `fallback_font` temporary mapping which is just annoying. I think lifetimes should allow for this (see my test: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7c84ae252dc58c23deaa6827e8072542).

Reply:
>But I think we can do way better than that by using the abstraction magic of methods. Say we move the check for the glyph index in the get_fallback_font method. We could easily have an early return if the loaded font already contains the glyph, which would probably clean things up a lot. We also wouldn't need the fallback_font temporary mapping which is just annoying. I think lifetimes should allow for this (see my test: 

I'm not sure about I understand your suggestion, since the fundamental problem here is that you create a new font, thus you don't have a reference to it at all. You can't return a reference to a stack local object from a method, like `get_fallback_font`, so this fallback should be stored **somewhere**, and I don't see where you'll store it with your approach.